### DISCUSSION TITLE
Submitting transactions using Pycardano

### DISCUSSION THREAD

**Discussion by while0x1** [OpShin/opshin#46] [Q&A] [answered]
https://preprod.cexplorer.io/tx/77b502b5b33318fb9b5524efdf7533f14e152d283e3a7b7d098393a99cd165f1

The datum subbmited in the above transaction renders as:
```json
{
    "fields": [ {
            "list": [ {
                    "fields": [ { "bytes": "f5b84180b5a3cca20052258fe....." },
                        {  "int": 20000000 } ],
                    "constructor": 0
                }  ]  } ], "constructor": 0
}
```
which was submitted to the contract as `winnerDetails = winnerList([winner(winner1hash,20000000)])`

Where winner list is
```python3
 @dataclass()
class winner(PlutusData):
    winnerhash: bytes
    lovelace: int
    
@dataclass()
class winnerList(PlutusData):
    winners: List[winner]
```
The contract compiles and uses the following function to downcast the Datum into the correct type 
```python3
def getWinnersDatum(txins: List[TxInInfo],token: Token) -> winnerList:
    winnersDatum = winnerList([winner(b"abc",0),winner(b"def",0)])
    for txi in txins:
        if txi.resolved.value.get(token.policy_id, {b"": 0}).get(token.token_name, 0) > 0:
            unionDatum = txi.resolved.datum
            if isinstance(unionDatum, SomeOutputDatum):    
                winnersDatum: winnerList = unionDatum.datum
                constructor = unionDatum.CONSTR_ID
    return winnersDatum
```
After being cast a for loop will iterate over the datum
```python3
  for i in win.winners:
              if i.winnerhash in ctx.tx_info.signatories:
                  allow_spend = i.lovelace
```
Trying to spend a UTXO @ the contract results in an error RawCborDecodeError -- Size mismatch when decoding TxAux.\\\\nExpected 2, but found 4.\\"),DecoderErrorDeserialiseFailure \\"Shelley Tx\\" (DeserialiseFailure 1 \\"Size mismatch when decoding \\\\nRecord RecD.....

After casting are the custom class and the datum equivalent ? 
What is the best way to debugg these types of errors in eopsin - can the eopsin eval command be used to verify the down cast of type?

**Comment by nielstron** [comment]
This doesn't sound like a downcasting issue. It sounds like an error much earlier in the evaluation - the node trying to deserialize either the contract you passed or the datum, either of which are not passed in the correct format. Please add the blockfrost/pycardano code trying to spend the transaction, otherwise it is impossible to find the issue.

**Reply by while0x1** [reply]
I have tried a couple of ways - one using pycardano's transaction builder and one using raw - both end up with the same error message. 

the script for spending the locked funds 
```
import cbor2
from blockfrost import ApiUrls
from pycardano import *

context = BlockFrostChainContext("preprod", base_url=ApiUrls.preprod.value)

# This artifact was generated in step 2
with open("raffle/script.cbor", "r") as f:
    script_hex = f.read()
#dummy
gift_script = PlutusV2Script(bytes.fromhex(script_hex))
script_hash = plutus_script_hash(gift_script)
network = Network.TESTNET
script_address = Address(script_hash, network=network)

payment_vkey = PaymentVerificationKey.load("payment.vkey")
payment_skey = PaymentSigningKey.load("payment.skey")

prize = 20000000
total_ada = 1500000 + 5000000 + prize
FEE = 1000000

contract_address = 'addr_test1wpq74hu0qdntr89320lhrsej0wqzgdv56faxjqn0uujlgxcxvvdfu'
supply_address = 'addr_test1vr6mssvqkk3uegsq2gjcleen9rrfl78q6stsj5zkdqgpmdcz3x4jh'

#NFT_TIX_UTXO & 1.5 ADA
tx_id1 = '77b502b5b33318fb9b5524efdf7533f14e152d283e3a7b7d098393a99cd165f1'
tx_in1 = TransactionInput.from_primitive([tx_id1, 1])
#ADA_UTXO 5ADA
tx_id2 = 'bdcd64a1478dc7d6a3ee1316eea7cf09dea9fff6d0d8c4f5e9d005d976adc43c'
tx_in2 = TransactionInput.from_primitive([tx_id2, 0])
#collatoral
col_input = TransactionInput.from_primitive([tx_id2, 1])
#scriptUTXOS
script_utxo = ['cb5acf7711fec0198254dbae3057111796fcf6c898e9e6b27a00c9cc8de54501',0]
tx_in_script = TransactionInput.from_primitive([script_utxo[0], script_utxo[1]])

policy_id = b"\xe7\x7b\x78\xf1\x37\xcf\x18\xf7\x13\x29\x04\x21\x85\x95\x99\x7f\xfa\x13\x2a\xb5\xa8\xd4\xc8\x32\xc3\xf1\xe3\x15" 
adaout1 =  1500000

asset_out = Value.from_primitive([adaout1,{policy_id: { b"VeriFair_alpha_ticket": 1}},])
asset_ret = Value.from_primitive([adaout1,{policy_id: { b"VeriFair_alpha_ticket": 2}},])

addr = Address.decode(contract_address)
nft_out = TransactionOutput(addr, asset_out)
addr = Address.decode(supply_address)
tix_back = TransactionOutput(addr, asset_ret)

changeADA = total_ada - FEE - adaout1 - adaout1

change = TransactionOutput(addr, changeADA,script=gift_script)

tx_body = TransactionBody(inputs=[tx_in1,tx_in2,tx_in_script], outputs=[nft_out, tix_back, change], fee=FEE, collateral=col_input)
signature = payment_skey.sign(tx_body.hash())
vk_witnesses = [VerificationKeyWitness(payment_vkey, signature)]
signed_tx = Transaction(tx_body, TransactionWitnessSet(vkey_witnesses=vk_witnesses))
context.submit_tx(signed_tx.to_cbor())
tx_body.id
```

**Reply by while0x1** [reply]
This yieleded a slightly different error than through transaction builder.. the raw error yieleded

```
error': 'Bad Request', 'message': '"transaction read error RawCborDecodeError [DecoderErrorDeserialiseFailure \\"Byron Tx\\" (DeserialiseFailure 1 \\"Size mismatch when decoding TxAux.\\\\nExpected 2, but found 4.\\"),DecoderErrorDeserialiseFailure \\"Shelley Tx\\" (DeserialiseFailure 1 \\"Size mismatch when decoding \\\\nRecord RecD.\\\\nExpected 4, but found 3.\\"),DecoderErrorDeserialiseFailure \\"Shelley Tx\\" (DeserialiseFailure 1 \\"Size mismatch when decoding \\\\nRecord RecD.\\\\nExpected 4, but found 3.\\"),DecoderErrorDeserialiseFailure \\"Shelley Tx\\" (DeserialiseFailure 1 \\"Size mismatch when decoding \\\\nRecord RecD.\\\\nExpected 4, but found 3.\\"),DecoderErrorDeserialiseFailure \\"Shelley Tx\\" (DeserialiseFailure 300 \\"expected list len or indef\\"),DecoderErrorDeserialiseFailure \\"Shelley Tx\\" (DeserialiseFailure 9186 \\"expected list len or indef\\")]"'
```

Im not sure what any of them really mean but the Exepcted list len or indef is new ..

**Reply by nielstron** [reply]
Might actually be an issue with the cardano-submit-api (which is used by @blockfrost) https://github.com/input-output-hk/cardano-node/issues/4829

You might have better luck using ogmios as a backend

**Reply by reitcircles** [reply]
> I have tried a couple of ways - one using pycardano's transaction builder and one using raw - both end up with the same error message.
> 
> the script for spending the locked funds
> 
> ```
> import cbor2
> from blockfrost import ApiUrls
> from pycardano import *
> 
> context = BlockFrostChainContext("preprod", base_url=ApiUrls.preprod.value)
> 
> # This artifact was generated in step 2
> with open("raffle/script.cbor", "r") as f:
>     script_hex = f.read()
> #dummy
> gift_script = PlutusV2Script(bytes.fromhex(script_hex))
> script_hash = plutus_script_hash(gift_script)
> network = Network.TESTNET
> script_address = Address(script_hash, network=network)
> 
> payment_vkey = PaymentVerificationKey.load("payment.vkey")
> payment_skey = PaymentSigningKey.load("payment.skey")
> 
> prize = 20000000
> total_ada = 1500000 + 5000000 + prize
> FEE = 1000000
> 
> contract_address = 'addr_test1wpq74hu0qdntr89320lhrsej0wqzgdv56faxjqn0uujlgxcxvvdfu'
> supply_address = 'addr_test1vr6mssvqkk3uegsq2gjcleen9rrfl78q6stsj5zkdqgpmdcz3x4jh'
> 
> #NFT_TIX_UTXO & 1.5 ADA
> tx_id1 = '77b502b5b33318fb9b5524efdf7533f14e152d283e3a7b7d098393a99cd165f1'
> tx_in1 = TransactionInput.from_primitive([tx_id1, 1])
> #ADA_UTXO 5ADA
> tx_id2 = 'bdcd64a1478dc7d6a3ee1316eea7cf09dea9fff6d0d8c4f5e9d005d976adc43c'
> tx_in2 = TransactionInput.from_primitive([tx_id2, 0])
> #collatoral
> col_input = TransactionInput.from_primitive([tx_id2, 1])
> #scriptUTXOS
> script_utxo = ['cb5acf7711fec0198254dbae3057111796fcf6c898e9e6b27a00c9cc8de54501',0]
> tx_in_script = TransactionInput.from_primitive([script_utxo[0], script_utxo[1]])
> 
> policy_id = b"\xe7\x7b\x78\xf1\x37\xcf\x18\xf7\x13\x29\x04\x21\x85\x95\x99\x7f\xfa\x13\x2a\xb5\xa8\xd4\xc8\x32\xc3\xf1\xe3\x15" 
> adaout1 =  1500000
> 
> asset_out = Value.from_primitive([adaout1,{policy_id: { b"VeriFair_alpha_ticket": 1}},])
> asset_ret = Value.from_primitive([adaout1,{policy_id: { b"VeriFair_alpha_ticket": 2}},])
> 
> addr = Address.decode(contract_address)
> nft_out = TransactionOutput(addr, asset_out)
> addr = Address.decode(supply_address)
> tix_back = TransactionOutput(addr, asset_ret)
> 
> changeADA = total_ada - FEE - adaout1 - adaout1
> 
> change = TransactionOutput(addr, changeADA,script=gift_script)
> 
> tx_body = TransactionBody(inputs=[tx_in1,tx_in2,tx_in_script], outputs=[nft_out, tix_back, change], fee=FEE, collateral=col_input)
> signature = payment_skey.sign(tx_body.hash())
> vk_witnesses = [VerificationKeyWitness(payment_vkey, signature)]
> signed_tx = Transaction(tx_body, TransactionWitnessSet(vkey_witnesses=vk_witnesses))
> context.submit_tx(signed_tx.to_cbor())
> tx_body.id
> ```

I had a similar issue. Whenever you use string address do convert like this before including in transaction.

Change:
```
addr = Address.decode(contract_address)  
```
to

 ```
 contract_address = 'addr_test1wpq74hu0qdntr89320lhrsej0wqzgdv56faxjqn0uujlgxcxvvdfu'
  caddr = Address.from_primitive(contract_address)
```

**Comment by nielstron** [comment, marked as answer]
It looks like this is an issue with the cardano-submit-api (which is used by @blockfrost) 

A related issue has been raised here: https://github.com/input-output-hk/cardano-node/issues/4829

You might want to try instead using the Ogmios backend offered by @Python-Cardano

**Reply by while0x1** [reply]
Thanks Neilstron :) -- after alot of juggling I was able to spin up a demeter.run preprod python environment - then use the eopsin scripts to create a cardano-cli plutus file - then after submitting the contract through CLI it executed and even returned the error message!

I realize now that: 
```
def win_policy_present(txins: List[TxInInfo],token: Token) -> bool:
    res = False
    for txi in txins:
        if txi.resolved.value.get(token.policy_id, {b"": 0}).get(token.token_name, 0) > 0:
            res = True
    return res
```
Fails because I sent it in a previous transaction .. but hopefully i can just include this utxo in the transaction and send it back to the contract.. :)

**Comment by nielstron** [comment]
This might also be an issue with using the wrong addresses. A common mistake during submission and its resolution was posted here:https://github.com/Python-Cardano/pycardano/issues/165#issuecomment-1433698213
